charms/interface_tokens/provides.py

```diff
@@ -0,0 +1,67 @@
+import json
+from dataclasses import dataclass
+from typing import List, Optional
+
+from ops import CharmBase, Relation, Unit
+
+
+@dataclass
+class TokenRequest:
+    """Represents a request for an authentication token.
+
+    Attributes:
+        unit (str): The name of the unit making the request.
+        user (str): The user associated with the token request.
+        group (str): The group associated with the user.
+    """
+
+    unit: str
+    user: str
+    group: str
+
+
+class TokensProvider:
+    """Implements the Provides side of the tokens interface."""
+
+    def __init__(self, charm: CharmBase, endpoint: str):
+        self.charm = charm
+        self.endpoint = endpoint
+
+    def send_token(self, request: TokenRequest, token: str) -> None:
+        """Send the token to the related units."""
+        # Aggregate tokens from all relations.
+        tokens = {
+            k: v
+            for relation in self.relations
+            for k, v in json.loads(relation.data[self.unit].get("tokens", "{}")).items()
+        }
+
+        tokens[request.user] = token
+        value = json.dumps(tokens)
+
+        # Update all relations with the new token data.
+        for relation in self.relations:
+            relation.data[self.unit]["tokens"] = value
```

Comment:
if you end up keeping track of the relation id... you'll filter here like this:

```suggestion
        for relation in self.relations:
            if request.id == relation.id:
                relation.data[self.unit]["tokens"] = value
```